### PR TITLE
add subscriber handler and status displays

### PR DESCRIPTION
**Pull request by ChristophTobler**

### CODE REVIEW

File: src/modules/events/status_display.h [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+#include <drivers/drv_hrt.h>
+
+#include <uORB/uORB.h>
+#include <uORB/topics/battery_status.h>
+#include <uORB/topics/cpuload.h>
+#include <uORB/topics/led_control.h>
+#include <uORB/topics/vehicle_status.h>
+#include <uORB/topics/vehicle_status_flags.h>
+
+#define CPU_OVERLOAD_VALUE 0.80f
```
Comment (bkueng) [inline src/modules/events/status_display.h]: These defines are unused, can you remove them?

Comment (ChristophTobler) [reply src/modules/events/status_display.h]: done

**Comment by dagar** [comment]
Remove mpu9250?

File: src/modules/events/status_display.cpp [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+		orb_copy(ORB_ID(vehicle_status_flags), _subscriber_handler.get_vehicle_status_flags_sub(), &_vehicle_status_flags);
+		got_updates = true;
+	}
+
+	if (_subscriber_handler.vehicle_status_updated()) {
+		orb_copy(ORB_ID(vehicle_status), _subscriber_handler.get_vehicle_status_sub(), &_vehicle_status);
+		got_updates = true;
+	}
+
+	if (_subscriber_handler.vehicle_attitude_updated()) {
```
Comment (dagar) [inline src/modules/events/status_display.cpp]: Attitude?

**Comment by dagar** [comment]
The semaphoreci failure is real (out of flash on px4fmu-v2), the circleci failure is intermittent and incredibly annoying. Let me know if you have any ideas for why it sporadically gets stuck in a reconfigure loop.

File: src/modules/events/status_display.cpp (hunk 2 above)
Comment (bkueng) [reply src/modules/events/status_display.cpp]: I'd keep it so that it's consistent with the topic name.

**Comment by bkueng** [comment]
Ok I disabled `mpu9250`.

@dagar there's some build dependency missing for incremental builds: disabling the driver and building the target gave me:
```
NuttX/apps/libapps.a(builtin_list.o):(.rodata.g_builtins+0x20c): undefined reference to `mpu9250_main'
```
A clean build fixed it. The other direction is even more problematic: enabling something gets silently ignored.

**Comment by dagar** [comment]
Yes I was aware of the nuttx builtins dependency issue, and working on a fix in this large PR. https://github.com/PX4/Firmware/pull/8043

I should get a minimal fix for that particular issue into v1.7.0.

File: src/modules/events/status_display.cpp (hunk 2 above)
Comment (dagar) [reply src/modules/events/status_display.cpp]: I meant why is attitude needed for status updates? I don't see it being used.

**Comment by dagar** [comment]
Looks good, one idea I've had for other areas is to update Subscription (https://github.com/PX4/Firmware/blob/master/src/modules/uORB/Subscription.hpp#L131) for use in the work queues. Either lazy subscribe or change the structure to allocate the entire class in the first cycle.

Another idea that could be useful in a few places is a variation that returned a copy of the latest data, but didn't hold an internal copy.

Feel free to ignore for this PR, just something to think about for later.

**Comment by bkueng** [comment]
>  change the structure to allocate the entire class in the first cycle.

This is already the case for the events module. In general, the new `ModuleBase` accounts for that and handles it better.

> Another idea that could be useful in a few places is a variation that returned a copy of the latest data, but didn't hold an internal copy.

Isn't `SubscriptionBase` pretty much that? I like the interface of `Subscription`, but it has IMO too much overhead: it inherits from a linked list, stores the metadata & instance (https://github.com/PX4/Firmware/blob/master/src/modules/uORB/Subscription.hpp#L87), and in addition a copy of the data.

**Comment by dagar** [comment]
> This is already the case for the events module. In general, the new ModuleBase accounts for that and handles it better.
Great! I was thinking of what the land_detector does and not completely aware of how ModuleBase is implemented.

It would be great to direct more effort into improving core building blocks like Subscription, Publication, BlockParam, etc rather than having the many different flavours of boilerplate that keep evolving.

**Comment by dagar** [comment]
What's the overall plan here? As commander will still need to subscribe and act on most (all?) of these topics would it make sense to publish something like an overall commander status or higher level LED commands (battery low, battery critical, etc)?

File: src/modules/events/status_display.cpp (hunk 2 above)
Comment (bkueng) [reply src/modules/events/status_display.cpp]: Alright, removed

**Comment by bkueng** [comment]
> As commander will still need to subscribe and act on most (all?) of these topics would it make sense to publish something like an overall commander status or higher level LED commands (battery low, battery critical, etc)?

Yes that makes sense. We just need to be careful to avoid exposing commander internals.

**Comment by ChristophTobler** [comment]
@bkueng Good to go from my side

**Comment by bkueng** [comment]
CI failure is unrelated

**Comment by dagar** [comment]
> Yes that makes sense. We just need to be careful to avoid exposing commander internals.

That's really the point I was trying to make. For example vehicle_status_flags are basically commander internals, but published for logging.

I also want to be careful with things like the battery thresholds. It's a trivial example, but in my opinion it would be better to have only a single module interpreting them and deciding if the battery is low. What if the battery percentage dips below the threshold, then goes above again? We can now have multiple power bricks with a battery status coming from each. How the low battery params are interpreted in that situation hasn't really been discussed in detail.

**Pull request merged**